### PR TITLE
Efficient conj

### PR DESCRIPTION
**Pull request by danielturek**
Removing all runtie `declare()` statements from conjugate samplers, and instead using persistent variables (and memory) declared in the `setup` function.

Addressing #507.

### CODE REVIEW

**Comment by fritzo** [comment]
Pro tip: if you change "Addressing #507" to "Fixes #507" then github will magically close #507 when this PR is merged.

Also, I wish we had some MCMC benchmarks set up to be able to see how much this effects runtime. Have you timed `test-mcmc.R` before and after this PR? If not, we can see this in travis logs when your tests complete.

File: packages/nimble/R/MCMC_conjugacy.R [hunk 1]
```diff
@@ -406,35 +409,36 @@ conjugacyClass <- setRefClass(
...
-                if(!cc_nodeInExpr(targetNode, linearityCheckExpr))                return(NULL)
-                if(cc_vectorizedComponentCheck(targetNode, linearityCheckExpr))   return(NULL)   # if targetNode is vectorized, make sure non of it's components appear in expr
-                linearityCheck <- cc_checkLinearity(linearityCheckExpr, targetNode)   # determines whether paramExpr is linear in targetNode
-                if(!cc_linkCheck(linearityCheck, link))                           return(NULL)
-                if(!cc_otherParamsCheck(model, depNode, targetNode))              return(NULL)   # ensure targetNode appears in only *one* depNode parameter expression
-                control <- addDependentNodeToControl(control, depNodeDist, depNode)
-            }
-            return(list(type=samplerType, target=targetNode, control=control))   # all dependent nodes passed the conjugacy check
-        },
+        ## trying commenting it out...  -DT July 2017
```
Comment (fritzo) [inline packages/nimble/R/MCMC_conjugacy.R]: nit: Please remove commented code before merging into devel.

File: packages/nimble/R/MCMC_conjugacy.R [hunk 2]
```diff
@@ -495,34 +495,40 @@ conjugacyClass <- setRefClass(
...
-                ##     N_DEP <- length(control$DEP_CONTROL_NAME)
-                ##     DEP_NODEFUNCTIONS <- nimbleFunctionList(NF_VIRTUAL)
-                ##     for(iDep in 1:N_DEP)
-                ##         DEP_NODEFUNCTIONS[[iDep]] <- model$nodes[[control$DEP_CONTROL_NAME[iDep]]]
-                ## }, list(N_DEP             = as.name(paste0('N_dep_', distName)),
-                ##         DEP_CONTROL_NAME  = as.name(paste0(  'dep_', distName)),
-                ##         DEP_NODEFUNCTIONS = as.name(paste0(  'dep_', distName, '_nfs')),
-                ##         NF_VIRTUAL        = as.name(paste0('node_stoch_', distName))))
+                
+                ## declare() statements are removed from run() code,
```
Comment (fritzo) [inline packages/nimble/R/MCMC_conjugacy.R]: Remove commented code before merging to devel.

File: packages/nimble/R/MCMC_conjugacy.R [hunk 3]
```diff
@@ -495,34 +495,40 @@ conjugacyClass <- setRefClass(
...
+                    depNodeParamNdim <- getDimension(distName, param)
+                    functionBody$addCode(DEP_PARAM_VAR <- array(0, dim = DECLARE_SIZE),
+                                         list(DEP_PARAM_VAR = as.name(paste0('dep_', distName, '_', param)),
+                                              DECLARE_SIZE  = makeDeclareSizeField(as.name(paste0('N_dep_', distName)), as.name(paste0('dep_', distName, '_nodeSizeMax')), as.name(paste0('dep_', distName, '_nodeSizeMax')), depNodeParamNdim)))
+                }
             }
-
+            
+            ## more new array() setup outputs, instead of declare() statements, for offset and coeff variables
+            ## July 2017
```
Comment (fritzo) [inline packages/nimble/R/MCMC_conjugacy.R]: Remove "July 2017". Use git blame or github blame view instead.

**Comment by danielturek** [comment]
Thanks for the feedback @fritzo

File: packages/nimble/R/MCMC_conjugacy.R (hunk 3 above)
Comment (fritzo) [inline packages/nimble/R/MCMC_conjugacy.R]: Rational for `git blame`: Consider two strategies: manually add historical comments to a select section of code, or rely entirely on automatic history annotation. While the latter strategy is less customizable, it has the major advantage that *every line of code* has history annotations. And while there is initial investment cost in gaining proficiency with `git blame` and `git grep`, this quickly pays off in saved time for code writers, and more focused attention for code readers (who more often than not are uninterested in history).

**Comment by danielturek** [comment]
Passed tests; requesting a quick review from @perrydv, before I merge this in.

**Comment by danielturek** [comment]
@perrydv Please let me know if you're able to look this over, I'm keen to merge it in.

**Comment by perrydv** [comment]
👍

**Comment by danielturek** [comment]
@perrydv thanks!

**Pull request merged**